### PR TITLE
spread instances equally to all AZs

### PR DESCRIPTION
**Pull request by zetaab**
fixes issue #6533 

Before:
screenshot 2019-02-26 at 9 25 09

After:
screenshot 2019-02-26 at 10 18 08

/sig openstack

/cc @drekle please review

### CODE REVIEW

File: pkg/model/openstackmodel/servergroup.go [hunk 1]
```diff
@@ -73,6 +73,15 @@ func (b *ServerGroupModelBuilder) buildInstances(c *fi.ModelBuilderContext, sg *
...
 	}
 
+	uniqueZones := make(map[string]string)
+	for _, subnet := range b.Cluster.Spec.Subnets {
+		uniqueZones[subnet.Zone] = ""
+	}
+	var zones []string
+	for k, _ := range uniqueZones {
+		zones = append(zones, k)
+	}
```
Comment (zetaab) [inline pkg/model/openstackmodel/servergroup.go]: is there simpler way to do this? So I want unique zone values from subnet zones

Comment (KashifSaadat) [reply pkg/model/openstackmodel/servergroup.go]: Have a peek at the maps & slice util packages in the repo, you could reuse or extend the functionality there to make this a little cleaner.

Comment (zetaab) [reply pkg/model/openstackmodel/servergroup.go]: @KashifSaadat in my opinion its now cleaner :)

File: pkg/model/openstackmodel/servergroup.go [hunk 2]
```diff
@@ -96,17 +104,27 @@ func (b *ServerGroupModelBuilder) buildInstances(c *fi.ModelBuilderContext, sg *
 		}
 		c.AddTask(portTask)
 
+		var az *string
+		if ig.Spec.Role == kops.InstanceGroupRoleMaster {
+			// take zone name from ig name, kind of hack. Is there easier way to see az to igs?
```
Comment (drekle) [inline pkg/model/openstackmodel/servergroup.go]: The `ig.Spec` has member Zones.

File: pkg/model/openstackmodel/servergroup.go [hunk 3]
```diff
@@ -96,17 +104,27 @@ func (b *ServerGroupModelBuilder) buildInstances(c *fi.ModelBuilderContext, sg *
...
+			az = fi.String(strings.Replace(ig.Name, "master-", "", 1))
+		} else {
+			// take index from name, and spread equally to zones
+			az = fi.String(zones[int(i)%len(zones)])
+		}
+
 		instanceTask := &openstacktasks.Instance{
```
Comment (drekle) [inline pkg/model/openstackmodel/servergroup.go]: we have an index to use to mod over all zones.  Would something like this work:

```
		zone := ig.Spec.Zones[ i % len(ig.Spec.Zones)]

		instanceTask := &openstacktasks.Instance{
			Name:        instanceName,
			Region:      fi.String(b.Cluster.Spec.Subnets[0].Region),
			Flavor:      fi.String(ig.Spec.MachineType),
			Image:       fi.String(ig.Spec.Image),
			SSHKey:      fi.String(sshKeyName),
```

Comment (zetaab) [reply pkg/model/openstackmodel/servergroup.go]: It does not work. ig.Spec.Zones is empty for all others than GCP

File: pkg/model/openstackmodel/servergroup.go (hunk 2 above)
Comment (zetaab) [reply pkg/model/openstackmodel/servergroup.go]: it has but its empty

File: pkg/model/openstackmodel/servergroup.go (hunk 3 above)
Comment (zetaab) [reply pkg/model/openstackmodel/servergroup.go]: see https://github.com/kubernetes/kops/blob/master/cmd/kops/create_cluster.go#L784

Comment (zetaab) [reply pkg/model/openstackmodel/servergroup.go]: ofc we could add openstack there as well?

**Comment by zetaab** [comment]
@drekle can you check now again? Of course we could add zone information to own struct. However, in case of openstack this information is already in place in subnets.

**Pull request merged**